proposal/forward-with.md

```diff
@@ -0,0 +1,174 @@
+// middleware.$saturation == 70%
+```
+
+Keyword arguments in the configuration must reference variable names as
+defined in the forwarded module, regardless of any concurent `as` clause:
+
+```scss
+// _origin.scss
+$hue: 0 !default;
+$color-hex: #ccc !default;
+```
+
+```scss
+// _middleware.scss
+@forward "origin" as color-* with (
+  $hue: 330, // the color-* prefix is not referenced in configuration
+  $color-hex: #966
+);
+```
+
+```scss
+// entrypoint.scss
+@use "middleware" as m;
+// m.$color-hue == 330
+// m.$color-hex == #966
+```
+
+A `@forward` rule configuration is applied to the source module even if the
+forwarding module acts as an entrypoint:
+
+```scss
+// _origin.scss
+$hue: 0 !default;
+```
+
+```scss
+// entrypoint.scss
+@forward "origin" with (
+  $hue: 330 !default
+);
+
+@use "origin"; // origin.$hue == 330
+```
+
+Multiple configurations can be chained in a single cascading "thread" that
+contains zero or more `@forward` rules, and zero or one terminal `@use` rule.
+Variables remain open to configuration in the chain as long as every mention
+includes the `!default` flag. Multiple threads configuring a single module will
+cause an error, even if they originate in the same file.
+
+## Syntax
+
+The new `WithClause` extends `@forward` to the follow grammar:
+
+<x><pre>
+**ForwardRule**     ::= '@forward' QuotedString AsClause? (ShowClause | HideClause)?  WithClause?
+**WithClause**      ::= 'with' '('
+&#32;                     KeywordArgument (',' KeywordArgument)\* ','?
+&#32;                   ')'
+**KeywordArgument** ::= '$' Identifier ':' Expression ('!default')
+</pre></x>
+
+## Semantics
+
+The `@forward ... with` semantics builds on the existing proposal for
+[Executing Files][], and should be understood as modifying and expanding upon
+the existing execution process rather than being a comprehensive replacement.
+
+[Executing Files]: ../accepted/module-system.md#executing-files
+
+Given a source file `file`, a configuration `config`, and an import context
+`import`:
+
+* When a `@forward` rule `rule` is encountered:
+
+  * For each variable `variable` in `config`:
+
+    * Let `rule-config` be an empty configuration.
+
+    * If `rule` has an `AsClause` with identifier `prefix`:
+
+      * If `variable`'s name begins with `prefix`, let `name` be the portion of
+        `variable`'s name after `prefix`.
+
+    * Otherwise, let `name` be `variable`'s name.
+
+    * Add a variable to `rule-config` with the name `name` and with the
+      same value as `variable`.
+
+  * Otherwise, let `rule-config` be `config`.
+
+  * If `rule` has a `WithClause`:
+
+    * For each `KeywordArgument` `argument` in this clause:
+
+      * Let `value` be the result of evaluating `argument`'s expression.
+
+      * If a variable exists in `rule-config` with the same name as `argument`'s
+        identifier:
+
+        * If `argument` does has a `!default` flag, do nothing.
+
+        * Otherwise, throw an error.
```

Comment:
Currently, all the error-checking for passing an invalid config variable is handled at the beginning of [Executing Files](https://github.com/sass/sass/blob/master/accepted/module-system.md#executing-files). We probably want to handle this in the same place; if we just error here, we'll end up forbidding this:

```scss
// downstream.scss
@use "upstream" with ($value: downstream);
```

```scss
// _midstream.scss
@forward "upstream" with ($value: midstream 1);

// Because midstream declares its own `$value` variable, it's valid for
// downstream to override it even though it's not overriding upstream's.
$value: midstream 2 !default;
```

This makes me realize that there's probably a better way to do that error checking, though. Instead of having it occur at the beginning of the execution of the file and basically re-writing the logic for figuring out which names are valid, we should put it in the `@use` clause execution block and have it check whether `config` contains any variables that aren't in `module`'s public API with a `!default` flag. Then you can use the same text to verify that `@forward ... with` doesn't provide extra config too, which I think currently isn't being done.

Reply:
I think I resolved the other issues. Have to run for now, but I'll dig into this one over the weekend.